### PR TITLE
sg: use 1pass instead of dev-private

### PR DESCRIPTION
**Pull request by jhchabran**
Drop dev-private entirely in favour of 1password. 

Fixes https://github.com/sourcegraph/sourcegraph/issues/29640 

- [ ] Find a better name for `sg secret download-config` 
- [ ] Provide a better error when running enteprise code without having ran the command
- Maybe plug the `requires_dev_private: true` field to call `sg secret download-config`'s code automatically ? 
## Test plan

- deleted `../dev-private` 
- ran `sg secret download-config` (with 1pass from `sg setup` configured) 
- ran `sg start` and got it working, with the right site-config and license.

### CODE REVIEW

File: dev/sg/internal/run/run.go [hunk 1]
```diff
@@ -346,6 +348,31 @@ var installFuncs = map[string]installFunc{
 		target := filepath.Join(root, fmt.Sprintf(".bin/docsite_%s", version))
 		return download.Executable(ctx, url, target)
 	},
+	"installEnterpriseFrontend": func(ctx context.Context, env map[string]string) error {
```
Comment (mrnugget) [inline dev/sg/internal/run/run.go:351]: Why "Frontend"?

File: dev/sg/internal/run/run.go [hunk 2]
```diff
@@ -385,20 +412,24 @@ func runWatch(
...
 			var cmdOut string
 			var err error
-			if cmd.Install != "" && cmd.InstallFunc == "" {
-				cmdOut, err = BashInRoot(ctx, cmd.Install, makeEnv(parentEnv, cmd.Env))
-			} else if cmd.Install == "" && cmd.InstallFunc != "" {
-				fn, ok := installFuncs[cmd.InstallFunc]
-				if !ok {
-					return errors.New("nope, no install func with that name")
+			secretsEnv, err := getSecrets(ctx, cmd)
+			if err == nil {
```
Comment (mrnugget) [inline dev/sg/internal/run/run.go:423]: I'd combine that with the `if err != nil {` below and have `if ... else `

**Review by mrnugget — COMMENTED** [review]
I like the idea, but I really, really hope that the workflow with 1pass is smooth (haven't tried it).

Also: what do we do about custom configurations? I personally only use a subset of the configuration in `dev-private`. We need to make sure that's still possible and that we don't tie people to the 1pass config. 

Another thing: how do we handle changes in the 1pass config? Before you had to explicitly do a `git pull` in dev-private to pull in new changes. Drawback of that was that folks missed new changes. Advantage: nothing messed with your local config while you weren't paying attention.

**Comment by bobheadxi** [comment]
> Also: what do we do about custom configurations? I personally only use a subset of the configuration in dev-private. We need to make sure that's still possible and that we don't tie people to the 1pass config.

The superset scenario can be covered by additional site config files but yeah, subset needs a workaround 🤔 

> Another thing: how do we handle changes in the 1pass config?

Cached external secrets are invalidated after a certain amount of time automatically, and can also easily be invalidated directly (https://github.com/sourcegraph/sourcegraph/pull/37034 has an example)

File: dev/sg/internal/run/run.go (hunk 1 above)
Comment (bobheadxi) [reply dev/sg/internal/run/run.go:351]: Frontend handles configuration loading IIRC so we "hijack" the installation here, is there another service that loads site config from disk?

Comment (mrnugget) [reply dev/sg/internal/run/run.go:351]: Ugh. Sorry @bobheadxi, I forgot to delete my comment there 😄 I thought that function would only install the configs and only later realised that it does install `enterprise-frontend`, so.. makes sense.

**Comment by mrnugget** [comment]
> Cached external secrets are invalidated after a certain amount of time automatically, and can also easily be invalidated directly (https://github.com/sourcegraph/sourcegraph/pull/37034 has an example)

Hmmm. That feels like it could catch you by surprise. What if we make "generate config" an explicit step? i.e.:

1. You run enterprise-frontend without the config it tells you "run `sg generate enterprise-config`" 
2. `sg setup` does that for you
3. if you run `enterprise-frontend` and sg detects that the config in 1pass has been updated, it tells you: "run `sg secrets ...` to review new changes, run `sg generate enterprise-config` to overwrite existing config"

**Comment by bobheadxi** [comment]
> What if we make "generate config" an explicit step?

We actually had this in the first draft (before https://github.com/sourcegraph/sourcegraph/pull/37035/commits/dd2ee9ae97b878cbf14a0b1401a3995effa3c6a6) but it felt rather clunky and surprising 😬  Also, with the explicit approach, the following would annoy you all the time if you ever make manual tweaks:

> if you run enterprise-frontend and sg detects that the config in 1pass has been updated, it tells you: "run sg secrets ... to review new changes, run sg generate enterprise-config to overwrite existing config"

**Comment by mrnugget** [comment]
Another idea: what if we use Go templating inside the default `.json` config files?

```json
{
  "GITHUB": [
    {
      "url": "https://github.com",
      //     HERE vvvv
      "token": "{{ 1pass-secret 'github-external-service' }}",
      "repositoryQuery": ["none"],
      "repos": [
        "sourcegraph-testing/go-repo-a",
      ]
    },
  ]
}
```

We could evaluate that when starting/installing enterprise-frontend and put the secrets in. That way we could commit the config and keep it in our repo, but the secrets wouldn't leak. People could then copy this file and overwrite the ENV var to use their own modifications.

And it's _visible_. Right now the content would be pulled from 1pass and put into a file without the user being able to inspect it, right?

**Comment by bobheadxi** [comment]
> Another idea: what if we use Go templating inside the default .json config files?

This sounds great actually! And we could populate it implicitly for you. It _could_ be tricky where if you're testing out a custom token or something it's easy to accidentally commit the token to the config though

> And it's visible. Right now the content would be pulled from 1pass and put into a file without the user being able to inspect it, right?

It's currently written to the new gitignore'd files that you can inspect, and the data is also available via `sg secret`:

```
enterprise/dev/site-config.json
enterprise/dev/external-services-config.json
```

**Comment by mrnugget** [comment]
> It's currently written to the new gitignore'd files that you can inspect

Yeah, I just realised that too. Hmm.

Another thought/data-point/whatever: reading through this function here https://github.com/sourcegraph/sourcegraph/pull/37035/files#diff-897504e4e746de30baca48b01357a93c150132b28960cc0e5575f760570a5756R356-R364 my thoughts were:

1. This will overwrite the file on every re-compile.
2. I can't put custom stuff in these files, because it would be overwritten.
3. Can I use `PRIVATE_SITECONFIG_DATA` and set it in `sg.config.overwrite.yaml`? Maybe, but that would be JSON inside YAML and that's always a pain in the butt
4. I'd have to use a custom JSON somewhere (where?) and set the EXTSVC_CONFIG path to it.

So that means while we have nice implicit defaults, if someone asks "how do I add another repository to my config?" we have to tell them "Create a new file, set an env var in sg.config.overwrite.yaml, copy the file contents of old file into new file, add repo", which means they end up completely opting out of the default behavior, just because they want to add 1 repository.

That's something I bet 90% of sourcegraph engineers do: modify list of repositories. And that means 90% of engineers would be opted-out of the default way.

So I'm wondering: how can we make this extensible?

**Comment by mrnugget** [comment]
@bobheadxi sorry for the huge braindump, but I started an "Edit:" in last comment and it went on longer than I thought it would 😄

**Comment by bobheadxi** [comment]
> So I'm wondering: how can we make this extensible?

+1 ! Note that we _do_ support multiple config files (https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/cli/config.go?L57-61), so for certain configurations you can just do an overwrite sets `SITE_CONFIG_FILE=enterprise/dev/site-config.json,...`. It doesn't look like we can do this for the external service stuff though: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/cli/config.go?L178:21&subtree=true . So... maybe we could? But I'm not sure how the merging behaviour should work (or how the site config merging works today)

I'll have to revisit this in the morning, maybe someone else will chime in before then 😛

**Comment by jhchabran** [comment]
> I think this is the question we need to answer, actually. It's not "how can we remove dev-private?", it's "how can we provide sane defaults, including secrets, and let users extend them easily?". 


Strong +1 too, the original initiative started with a `sg secret download-config` which preserved the original workflow, but kept exploring :)

**Comment by bobheadxi** [comment]
Morning thought: I think the templating approach is the way to go:

1. we can ship the template as a `go:embed`, and `sg generate config`  the following creates a template file _that is gitignored_
3. Change `DevPrivate` 1pass entry to a set of fields, rather than giant JSON
4. We populate template explicitly with secrets declared in `sg.config.yaml` at runtime, write it back to another file is also gitignored (this might be clunky, maybe the populated config could go in `/tmp` or similar?)

**Comment by mrnugget** [comment]
> We populate template explicitly with secrets declared in sg.config.yaml at runtime, write it back to another file is also gitignored (this might be clunky, maybe the populated config could go in /tmp or similar?)

Why not populate it when running `sg generate config`? Seems to be the most comfortable for me: you run `sg generate config` (I like that command name!) and end up with a file that's gitignored and in which the secrets are contained and you can modify it to whatever you want.

I'm a bit skeptical of silently populating secrets into the config at runtime, since, for example, I use my own tokens in the external config (and I bet a lot of others do too). So in order to keep using those and have a custom external services config, I'd have to `sg generate config` _and_ create an `sg.config.overwrite.yaml` with new secrets.

Sure, we could say that the user puts their own tokens in the file and that this is then skipped, but I'm wondering if a single step isn't simpler (and easier to understand and to eventually debug?).

**Comment by jhchabran** [comment]
> but I'm wondering if a single step isn't simpler (and easier to understand and to eventually debug?).

I'm for this as well, it gives our users a clear path to reset the state, by simply running the command again.

**Comment by mrnugget** [comment]
> I'm for this as well, it gives our users a clear path to reset the state, by simply running the command again.

Yes, good point! that's a good property, _but_ also something we should be aware of: `sg generate config` could maybe prompt the user to ask "do you want to overwrite this file?" and show a diff, so users don't accidentally overwrite their config in case `sg generate config` ever generates more than 1-2 files.

**Comment by jhchabran** [comment]
[Quick poll to gather some feedback.](https://sourcegraph.slack.com/archives/C01N83PS4TU/p1656408625186399) I find the experience of 1pass cli a bit annoying (having to login and eval stuff out the blue once in a while)

**Comment by jhchabran** [comment]
With our recent decision to drop 1pass integration, we can safely close this. We'll come back at this, maybe with GSM instead :)

**Pull request closed without merging**